### PR TITLE
add redis sentinel support

### PR DESCRIPTION
**Pull request by terencehonles**
This change adds an additional connection pool factory which can create Redis SentinelConnectionPool instances. Although prior work mainly contained the code in a specialized client, looking at redis-py and how it was written, this actually seemed like a more consistent implementation with regards to how this package implements its shared connection pool and the upstream sentinel support.

This code was inspired by the following:

- django-redis-sentinel [1] [2]
- django-redis-sentinel-redux [3]

[1]: https://github.com/KabbageInc/django-redis-sentinel
[2]: https://github.com/SpatialBuzz/django-redis-sentinel
[3]: https://github.com/danigosa/django-redis-sentinel-redux

fixes: #426

### CODE REVIEW

**Comment by qeternity** [comment]
What's holding this back? It seems with redis-py now having very accessible sentinel/cluster support that we ought to have this implemented.

**Comment by terencehonles** [comment]
> What's holding this back? It seems with redis-py now having very accessible sentinel/cluster support that we ought to have this implemented.

Probably a review and to see if this is actually the direction the project wants to support. To chime in about if it's working, we've been using this in production since the PR was made so we've not seen any problems with it.

**Comment by WisdomPill** [comment]
Hey there,
Thanks for contributing.

I have had a look into this but since I do not now much about redis sentinel I did not review this yet.
I read a little bit about it but I might be slow at responding into this specific issue.

Sorry, maybe there is somebody else that knows more about sentinel than me and can review it quicker

**Comment by WisdomPill** [comment]
@terencehonles can you please update the ci? now we have moved to GitHub actions

**Comment by terencehonles** [comment]
done

**Review by WisdomPill — CHANGES_REQUESTED** [review]

**Comment by WisdomPill** [comment]
I do not feel confident in reviewing the sentinel part, I do not know the corner cases, maybe there is a brave should that could step in? Or we can just count the votes on a comment to say that it is approved by a good amount of users?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -56,6 +64,34 @@ jobs:
 
     - name: Tox tests
       run: |
+        # setup a redis sentinel
```
[the remark this answers is not in the document]
Comment (terencehonles) [reply .github/workflows/test.yml:67]: you can't specify a command to a service as of now and this needs to be on the host network since redis does not use DNS names but IPs only. For the master this isn't required because clients will ignorantly connect to 127.0.0.1 but because of the port specification that will just work. For the sentinel 127.0.0.1 is itself which is not correct and it needs to be on the host network so that it gets the same port forwarding the client gets.

File: django_redis/client/sentinel.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from urllib.parse import parse_qs, urlencode, urlparse, urlunparse
+
+from django.core.exceptions import ImproperlyConfigured
+from redis.sentinel import SentinelConnectionPool
+
+from .default import DefaultClient
+
+
+def replace_query(url, query):
+    return urlunparse((*url[:4], urlencode(query, doseq=True), url[5]))
```
[the remark this answers is not in the document]
Comment (terencehonles) [reply django_redis/client/sentinel.py:10]: yes, schemes aren't handled specifically:

```python
>>> from urllib.parse import urlencode, urlparse, urlunparse
>>>
>>> urlparse('unix://test')
ParseResult(scheme='unix', netloc='test', path='', params='', query='', fragment='')
>>> urlparse('fake://test')
ParseResult(scheme='fake', netloc='test', path='', params='', query='', fragment='')
>>>
>>> url = urlparse('fake://test')
>>> query = dict(test=1, other=[1,2])
>>> urlunparse((*url[:4], urlencode(query, doseq=True), url[5]))
'fake://test?test=1&other=1&other=2'
```

Aside from them appearing in the URL and some basic handling in `django_redis/pool.py` they are just to create the kwargs for the underlying connection pool.

**Comment by AnesFoufa** [comment]
> I do not feel confident in reviewing the sentinel part, I do not know the corner cases, maybe there is a brave should that could step in? Or we can just count the votes on a comment to say that it is approved by a good amount of users?

I think it should be released with a warning. It is not supposed to alter existing features and if released, we will get feedback on corner cases.
By the way, all the existing Sentinel caches for django are not maintained.

**Comment by terencehonles** [comment]
I think that is fine with me. I've been using it in production without a problem, but getting more eyes on it would be helpful.

> By the way, all the existing Sentinel caches for django are not maintained.

@AnesFoufa can you expand on this? Or just saying there are none that are maintained so this one not being marked as officially supported is fine?

By warning is a warning in the documentation and possibly also a doc string sufficient?

File: django_redis/pool.py [hunk 3]
```diff
@@ -117,6 +121,53 @@ def get_connection_pool(self, params):
...
+        # provide the connection pool kwargs to the sentinel in case it
+        # needs to use the socket options for the sentinels themselves
+        connection_kwargs = self.make_connection_params(None)
+        connection_kwargs.pop("url")
+        connection_kwargs.update(self.pool_cls_kwargs)
+        self._sentinel = Sentinel(
+            sentinels,
+            sentinel_kwargs=options.get("SENTINEL_KWARGS"),
+            **connection_kwargs,
+        )
```
Comment (terencehonles) [inline django_redis/pool.py:147]: One thing I realized when reviewing https://github.com/jazzband/django-redis/pull/460/files#r556859461 for implications beyond "does the expected string get generated?" is that unlike the `ConnectionFactory` this will create a `Sentinel` object upon class creation. This _could_ be delayed to the first call to `get_connection_pool` if it was important not to construct a `Sentinel` object. It doesn't look like it would actually make a connection unless `single_connection_client` is `True`, but if py-redis changed its implementation this might make a connection when the factory is created rather when it's used.

**Comment by terencehonles** [comment]
Any thoughts or follow ups? @WisdomPill not sure what you were requesting me to change. It more looks like you wanted to ask a question about what I did rather than suggest something _needed_ changing. I believe my responses clarify what is happening and the "change requested" marker should at least be removed from the PR (unless there is a concrete change that should or could be made)

**Review by WisdomPill — APPROVED** [review]

File: django_redis/client/sentinel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WisdomPill) [reply django_redis/client/sentinel.py:10]: but this does not work with `rediss://` because it is 5 chars not 4

File: README.rst [hunk 4]
```diff
@@ -589,6 +589,87 @@ Django setting ``DJANGO_REDIS_CONNECTION_FACTORY``.
...
+    SENTINELS = [
+        ('sentinel-1', 26379),
+        ('sentinel-2', 26379),
+        ('sentinel-3', 26379),
+    ]
+
+    CACHES = {
+        "default": {
+            "BACKEND": "django_redis.cache.RedisCache",
+            "LOCATION": "redis://service_name/db",
```
Comment (severindellsperger) [inline README.rst:617]: Just wondering: why do we need a `LOCATION` here? 
I mean the current master should be identified from the sentinels, shouldn't it?

**Review by severindellsperger — COMMENTED** [review]
I just overflew your changes in the README.
I think if the mentioned points are fixed I can take a more detailed view of the code and implement it in my test setup.

File: README.rst (hunk 4 above)
Comment (terencehonles) [reply README.rst:617]: This specifies which replica set to ask for when talking to the sentinels. The request to the sentinels will be:

```
# primary
SENTINEL MASTER <service_name>
# replicas
SENTINEL REPLICAS <service_name>
```

I had left the `LOCATION` a normal redis URL because that seemed simpler and allows this code to use a lot of existing code.

File: README.rst [hunk 5]
```diff
@@ -589,6 +589,87 @@ Django setting ``DJANGO_REDIS_CONNECTION_FACTORY``.
...
+Use the sentinel connection factory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to facilitate using `Redis Sentinels`_, django-redis comes with a
+built in sentinel connection factory, which creates sentinel connection pools.
+In order to enable this functionality you should add the following:
+
+
+.. code-block:: python
+    :caption: myproj/settings.py
```
[the remark this answers is not in the document]
Comment (terencehonles) [reply README.rst:601]: Looking at https://github.com/github/markup it looks like this https://github.com/github/markup/blob/master/lib/github/commands/rest2html script is responsible, but it might be docutils' fault in the end (I haven't looked that far)

**Comment by WisdomPill** [comment]
Thanks to @severindellsperger for confirming that this PR is acceptable.

File: README.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (terencehonles) [reply README.rst:601]: I opened https://github.com/github/markup/pull/1452 to hopefully fix this. If not we should remove the Sphinx directives since I don't think any documentation is built with Sphinx anymore.
